src/core/thread/meshcop_dataset_manager.cpp

```diff
@@ -0,0 +1,544 @@
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements common methods for manipulating MeshCoP Datasets.
+ *
+ */
+
+#include <stdio.h>
+
+#include <coap/coap_header.hpp>
+#include <common/code_utils.hpp>
+#include <common/logging.hpp>
+#include <common/timer.hpp>
+#include <platform/random.h>
+#include <thread/meshcop_dataset.hpp>
+#include <thread/meshcop_dataset_manager.hpp>
+#include <thread/thread_netif.hpp>
+#include <thread/thread_tlvs.hpp>
+#include <thread/thread_uris.hpp>
+
+namespace Thread {
+namespace MeshCoP {
+
+DatasetManager::DatasetManager(ThreadNetif &aThreadNetif, const char *aUri):
+    mMle(aThreadNetif.GetMle()),
+    mNetif(aThreadNetif),
+    mNetworkDataLeader(aThreadNetif.GetNetworkDataLeader()),
+    mResource(aUri, &HandleSet, this),
+    mTimer(&HandleTimer, this),
+    mUri(aUri),
+    mCoapServer(aThreadNetif.GetCoapServer())
+{
+    mCoapServer.AddResource(mResource);
+}
+
+ThreadError DatasetManager::Set(const Dataset &aDataset, uint8_t &aFlags)
+{
+    ThreadError error = kThreadError_None;
+
+    aFlags = 0;
+
+    VerifyOrExit(mNetwork.GetTimestamp().Compare(aDataset.GetTimestamp()) > 0, error = kThreadError_InvalidArgs);
+
+    mLocal = aDataset;
+    aFlags |= kFlagLocalUpdated;
+
+    switch (mMle.GetDeviceState())
+    {
+    case Mle::kDeviceStateChild:
+    case Mle::kDeviceStateRouter:
+        if (mNetwork.GetTimestamp().Compare(mLocal.GetTimestamp()) > 0)
+        {
+            mTimer.Start(1000);
+        }
+
+        break;
+
+    case Mle::kDeviceStateLeader:
+        mNetwork = mLocal;
+        aFlags |= kFlagNetworkUpdated;
+        mNetworkDataLeader.IncrementVersion();
+        mNetworkDataLeader.IncrementStableVersion();
+        break;
+
+    default:
+        break;
+    }
+
+exit:
+    return error;
+}
+
+ThreadError DatasetManager::Set(const Timestamp &aTimestamp, const Message &aMessage,
+                                uint16_t aOffset, uint16_t aLength, uint8_t &aFlags)
+{
+    ThreadError error = kThreadError_None;
+    int compare;
+
+    aFlags = 0;
+
+    mNetwork.SetTimestamp(aTimestamp);
+    SuccessOrExit(error = mNetwork.Set(aMessage, aOffset, aLength));
+    aFlags |= kFlagNetworkUpdated;
+
+    compare = mLocal.GetTimestamp().Compare(aTimestamp);
+
+    if (compare > 0)
+    {
+        mLocal.SetTimestamp(aTimestamp);
+        SuccessOrExit(error = mLocal.Set(aMessage, aOffset, aLength));
+        aFlags |= kFlagLocalUpdated;
+    }
+    else if (compare < 0)
+    {
+        mTimer.Start(1000);
```

Comment:
Do we need to check whether mTimer is already running or not in case for the timer does not fire? Then the dataset may not be updated.